File: y2022-2023/AaSD20230619/A5.py
from random import randint
import math

lazy_leng = 1000
base_len = 3
base = 10 ** base_len


def norm_vec(a):
    for i in range(len(a) - 1):
        if a[i] < 0:
            a[i] = base + a[i]
            a[i + 1] -= 1
        a[i + 1] += a[i] // base
        a[i] %= base
    while a[-1] >= base:
        a.append(a[-1] // base)
        a[-2] %= base


def print_number(a):
    stri = ""
    for i in range(len(a)):
        if i != 0:
            num = a[-1 - i]
            if num == 0:
                num += 1
            len_num = int(math.log10(num)) + 1
            for j in range(base_len - len_num):
                stri += '0'
        stri += str(a[-1 - i])
    return stri


def print_part_res(a, b, dig):
    print_number(a)
    print(" " + dig + " ", end='')
    print_number(b)
    print()


def print_part_res_with_ans(a, b, c, dig):
    print_number(a)
    print(" " + dig + " ", end='')
    print_number(b)
    print(" = ", end='')
    print_number(c)
    print()


def min_nulls(a):
    k = 0
    for i in range(len(a)):
        if a[i] != 0:
            k = i + 1
    if k == 0:
        return [0]
    if k == len(a):
        return a
    b = []
    for i in range(k):
        b.append(a[i])
    return b


def add_nulls(a, b):
    while len(a) < len(b):
        a.append(0)
    while len(b) < len(a):
        b.append(0)


def sum_vec(a, b):
    add_nulls(a, b)
    c = [0] * len(a)
    for i in range(len(b)):
        c[i] = a[i] + b[i]
    norm_vec(c)
    return c


def sub_vec(a, b):
    add_nulls(a, b)
    c = [0] * len(a)
    for i in range(len(b)):
        c[i] = a[i] - b[i]
    norm_vec(c)
    return c


def mul_vec(a, b):
    add_nulls(a, b)
    c = [0] * (2 * len(a))
    if len(b) < lazy_leng:
        for i in range(len(a)):
            for j in range(len(b)):
                c[i + j] += a[i] * b[j]
    else:
        k = len(a) // 2
        a1, a2, b1, b2 = [], [], [], []
        for i in range(k):
            a1.append(a[i])
            b1.append(b[i])
        for i in range(k, len(a)):
            a2.append(a[i])
            b2.append(b[i])
        a1b1 = mul_vec(a1, b1)
        a2b2 = mul_vec(a2, b2)
        a1_a2 = sum_vec(a1, a2)
        b1_b2 = sum_vec(b1, b2)
        a1_a2b1_b2 = mul_vec(a1_a2, b1_b2)
        part_sub = sub_vec(a1_a2b1_b2, a1b1)
        a1b2_a2b1 = sub_vec(part_sub, a2b2)
        a1b1 = min_nulls(a1b1)
        a2b2 = min_nulls(a2b2)
        a1b2_a2b1 = min_nulls(a1b2_a2b1)
        for i in range(len(a1b1)):
            c[i] += a1b1[i]
        for i in range(2 * k, 2 * k + len(a2b2)):
            c[i] += a2b2[i - 2 * k]
        for i in range(k, k + len(a1b2_a2b1)):
            c[i] += a1b2_a2b1[i - k]
    return c


# v1 = input()
# v2 = input()
cnst = 10 ** 10
epoch = 10 ** 6
for i in range(epoch):
    aa = randint(-cnst, cnst)
    v1 = str(aa)
    bb = randint(-cnst, cnst)
    v2 = str(bb)
    less_zero = False
    if ((v1[0] == '-' and v2[0] != '-') or (v1[0] != '-' and v2[0] == '-')):
        less_zero = True
    leng = max(len(v1) - (1 if v1[0] == '-' else 0), len(v2) - (1 if v2[0] == '-' else 0))
    vec_len = ((leng + base_len) // base_len)
    val1 = [0] * vec_len
    val2 = [0] * vec_len
    k = 0
    for i in range(len(v1) - 1, -1, -base_len):
        a = 0
        for j in range(base_len - 1, -1, -1):
            if (i - j >= 0) and (v1[i - j] != '-'):
                a *= 10
                a += ord(v1[i - j]) - ord('0')
        val1[k] = a
        k += 1
    k = 0
    for i in range(len(v2) - 1, -1, -base_len):
        a = 0
        for j in range(base_len - 1, -1, -1):
            if (i - j >= 0) and (v2[i - j] != '-'):
                a *= 10
                a += ord(v2[i - j]) - ord('0')
        val2[k] = a
        k += 1
    # if less_zero:
    #     print('-')
    if(aa == 0 or bb == 0):
        print(0)
    ans = mul_vec(val1, val2)
    norm_vec(ans)
    ans = min_nulls(ans)
    l = print_number(ans)
    if len(l) == 1 and l == "0":
        less_zero = False
    if less_zero:
        l = '-' + l
    if str(aa * bb) != l:
        print(aa, bb, aa * bb, l)
        print()

File: y2022-2023/AaSD20230619/A5.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <ctime>

using namespace std;


typedef long long ll;
typedef unsigned long long ull;

ull lazy_leng = 10;
ll base_len = 5;
ll base;

void norm_vec(vector<ll> &a) {
    for (ull i = 0; i < a.size() - 1; i++) {
        if (a[i] < 0) {
            a[i] = base + a[i];
            a[i + 1] -= 1;
        }
        a[i + 1] += a[i] / base;
        a[i] %= base;
    }
    while (a[a.size() - 1] >= base) {
        a.push_back(a[a.size() - 1] / base);
        a[a.size() - 2] %= base;
    }
}

void print_number(vector<ll> &a) {
    for (ull i = 0; i < a.size(); i++) {
        if (i != 0) {
            ll num = a[a.size() - 1 - i];
            if (num == 0) {
                num += 1;
            }
            ll len_num = 0;
            while (num != 0) {
                len_num += 1;
                num /= 10;
            }
            for (ll j = 0; j < base_len - len_num; j++) {
                cout << '0';
            }
        }
        cout << a[a.size() - 1 - i];
    }
}

void print_part_res(vector<ll> &a, vector<ll> &b, char dig) {
    print_number(a);
    cout << " " << dig << " ";
    print_number(b);
    cout << endl;
}

void print_part_res_with_ans(vector<ll> &a, vector<ll> &b, vector<ll> &c, char dig) {
    print_number(a);
    cout << " " << dig << " ";
    print_number(b);
    cout << " = ";
    print_number(c);
    cout << endl;
}

vector<ll> min_nulls(vector<ll> &a) {
    ull k = 0;
    for (ull i = 0; i < a.size(); i++) {
        if (a[i] != 0) {
            k = i + 1;
        }
    }
    if (k == 0) {
        return vector<ll>{0};
    }
    if (k == a.size()) {
        return a;
    }
    vector<ll> b;
    for (ull i = 0; i < k; i++) {
        b.push_back(a[i]);
    }
    return b;
}

void add_nulls(vector<ll> &a, vector<ll> &b) {
    while (a.size() < b.size()) {
        a.push_back(0);
    }
    while (b.size() < a.size()) {
        b.push_back(0);
    }
}

vector<ll> sum_vec(vector<ll> &a, vector<ll> &b) {
    add_nulls(a, b);
    vector<ll> c(a.size());
    for (ull i = 0; i < b.size(); i++) {
        c[i] = a[i] + b[i];
    }
    norm_vec(c);
//    print_part_res_with_ans(a, b, c, '+');
    return c;
}


vector<ll> sub_vec(vector<ll> &a, vector<ll> &b) {
    add_nulls(a, b);
    vector<ll> c(a.size());
    for (ull i = 0; i < b.size(); i++) {
        c[i] = a[i] - b[i];
    }
    norm_vec(c);
//    print_part_res_with_ans(a, b, c, '-');
    return c;
}

vector<ll> mul_vec(vector<ll> &a, vector<ll> &b) {
    add_nulls(a, b);
    vector<ll> c(2 * a.size());
    if (b.size() < lazy_leng) {
        for (ull i = 0; i < a.size(); i++) {
            for (ull j = 0; j < b.size(); j++) {
                if(i + j > c.size()) {
                    cout << "f**k all" << endl;
                }
                c[i + j] += a[i] * b[j];
            }
        }

    } else {
        ull k = a.size() / 2;
        vector<ll> a1, a2, b1, b2;
        for (ull i = 0; i < k; i++) {
            a1.push_back(a[i]);
            b1.push_back(b[i]);
        }
        for (ull i = k; i < a.size(); i++) {
            a2.push_back(a[i]);
            b2.push_back(b[i]);
        }
        vector<ll> a1b1 = mul_vec(a1, b1);
        vector<ll> a2b2 = mul_vec(a2, b2);
        vector<ll> a1_a2 = sum_vec(a1, a2);
        vector<ll> b1_b2 = sum_vec(b1, b2);
        vector<ll> a1_a2b1_b2 = mul_vec(a1_a2, b1_b2);
        vector<ll> part_sub = sub_vec(a1_a2b1_b2, a1b1);
        vector<ll> a1b2_a2b1 = sub_vec(part_sub, a2b2);
        a1b1 = min_nulls(a1b1);
        a2b2 = min_nulls(a2b2);
        a1b2_a2b1 = min_nulls(a1b2_a2b1);
        for (ull i = 0; i < a1b1.size(); i++) {
            if(i > c.size()) {
                cout << "f**k you and me" << endl;
            }
            c[i] += a1b1[i];
        }
        for (ull i = 2 * k; i < 2 * k + a2b2.size(); i++) {
            if(i > c.size()) {
                cout << "f**k you" << endl;
            }
            c[i] += a2b2[i - 2 * k];
        }
        for (ull i = k; i < k + a1b2_a2b1.size(); i++) {
            if(i > c.size()) {
                cout << "f**k me" << endl;
            }
            c[i] += a1b2_a2b1[i - k];
        }
    }
    norm_vec(c);
//    print_part_res_with_ans(a, b, c, '*');
    return c;
}

string random_number(int min, int max) {
    ull len = min + rand() % (max - min);
    string s;
    for (ull i = 0; i < len; i++) {
        if (i == 0) {
            s += ((rand() % 9) + 1) + '0';
        } else {
            s += (rand() % 10) + '0';
        }
    }
    return s;
}

int main() {

    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);
    srand(time(nullptr));
    base = (ll) pow(10, base_len);
    string v1, v2;
    cin >> v1 >> v2;
    bool less_zero = false;
    if ((v1[0] == '-' && v2[0] != '-') || (v1[0] != '-' && v2[0] == '-')) {
        less_zero = true;
    }
    ull len1 = v1.size();
    if(v1[0] == '-') {
        len1--;
    }
    ull len2 = v2.size();
    if(v2[0] == '-') {
        len2--;
    }
    ull len = 250000 / base_len;
    vector<ll> val1(len), val2(len);
    ll k = 0;
    for (ll i = (ll) v1.size() - 1; i >= 0; i -= base_len) {
        ll a = 0;
        for (ll j = base_len - 1; j >= 0; j--) {
            if ((i - j >= 0) && (v1[i - j] != '-')) {
                a *= 10;
                a += v1[i - j] - '0';
            }
        }
        if(val1.size() == k) {
            cout << "f**k f**k" << endl;
        }
        val1[k] = a;
        k++;
    }
    k = 0;
    for (ll i = (ll) v2.size() - 1; i >= 0; i -= base_len) {
        ll a = 0;
        for (ll j = base_len - 1; j >= 0; j--) {
            if ((i - j >= 0) && (v2[i - j] != '-')) {
                a *= 10;
                a += v2[i - j] - '0';
            }
        }
        if(val2.size() == k) {
            cout << "f**k f**k" << endl;
        }
        val2[k] = a;
        k++;
    }
    vector<ll> ans = mul_vec(val1, val2);
    norm_vec(ans);
    ans = min_nulls(ans);
    if (ans.size() == 1 && ans[0] == 0) {
        less_zero = false;
    }
    if (less_zero) {
        cout << '-';
    }
    print_number(ans);
    cout << endl;
    return 0;
}